### ISSUE TITLE
reflect: gccgo and go differ on handling of fields from embedded structs

### ISSUE THREAD

**Issue by thinkofdeath** [golang/go#11007]
Test case: http://play.golang.org/p/pqIQCAJOXp

On go 1.4.2 and tip that works fine (as seen on the playground) but on gccgo 
`go version go1.4.2 gccgo (Ubuntu 5.1~rc1-0ubuntu1) 5.0.1 20150414 (prerelease) [gcc-5-branch revision 222102] linux/amd64`
it will fail with 
```
&{{}}
panic: reflect: reflect.mustBeAssignable.N13_reflect.Value using value obtained using unexported field

goroutine 16 [running]:
reflect.mustBeAssignable.N12_reflect.flag
	../../../src/libgo/go/reflect/value.go:216
reflect.mustBeAssignable.N13_reflect.Value
	../../../src/libgo/go/reflect/value.go:33
reflect.SetString.N13_reflect.Value
	../../../src/libgo/go/reflect/value.go:1373
main.main
	/home/matt/.local/share/data/liteide/liteide/goplay.go:19
created by main
	../../../src/libgo/runtime/go-main.c:42
```

This might be related to #7363 however I assumed that Exported fields from embedded structs should have the same visibility as if it was on the parent struct.

**Comment by minux** [comment]
yeah, gc and gccgo implemented different rules regarding
embedded structs.

gc seems to think all embedded structs are settable.
gccgo thinks all embedded structs are unsettable.

**Comment by thinkofdeath** [comment]
Seems to me that a mix between the two is the right option, matching the language.
e.g. 
```go
// package example
type unexported struct {
	Exported string
}

type Exported struct {
	unexported
}

// package main
import "example"

func main() {
    val := &example.Exported{}

    val.Exported = "test" // Ok
    reflect.ValueOf(val).Elem().FieldByName("Exported").SetString("Hello world") // Also ok

    val.unexported = /*something*/  // invalid
    reflect.ValueOf(val).Elem().Field(0).Set(/*something*/) // Also invalid
}
```

**Comment by ianlancetaylor** [comment]
Looks like #7363 .

**Comment by ianlancetaylor** [comment]
Oh, I see that you saw that.  Hmmm.  This does seem to be an issue in the reflect package.  It doesn't currently consider how an embedded field affects visibility.

**Comment by rsc** [comment]
The fix for #7363 should address this too.
